File: framework/python/src/soda/unittest/bootstrap.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().FUNCTION)

    # step [3]: setup other options
    # work.validator = (e,r) => bool
    work.compareSerial = True
    # work.setArgumentParser(index, (a) => b)
    # work.resultSerializer = (r) => s
    # work.resultParser = (s) => r
    work.run()

File: framework/java/src/main/java/soda/unittest/__Bootstrap__.java
package soda.unittest;

import java.util.*;
import java.util.function.Function;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.*;

class Solution {
    public int add(int a, int b) { return a + b; }
}

public class __Bootstrap__ implements Function<String, String> {

    @Override
    public String apply(String text) {
        // * create work by method, including method that has no return value
        var work = GenericTestWork.create(new Solution()::add);

        // * by class of data struct
        // var work = GenericTestWork.forStruct(Solution.class);

        // * for legacy code, discouraged
        // var work = TestWork.forObject(new Solution(), "METHOD");
        // var work = TestWork.forStruct(Struct.class);

        // * setup validator
        // work.setValidator((e, r) -> {...});
        work.setCompareSerial(true);
        return work.run(text);
    }

    public static void main(String[] args) throws Exception {
        System.out.println(new __Bootstrap__().apply(Utils.fromStdin()));
    }

}
